drivers/spi/spi_gd32.c

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright (c) 2021 BrainCo Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gd_gd32_spi
+
+#include <errno.h>
+#include <kernel.h>
+#include <drivers/pinctrl.h>
+#include <drivers/spi.h>
+#include <soc.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(spi_gd32);
+
+#include "spi_context.h"
+
+/* SPI error status mask. */
+#define SPI_GD32_ERR_MASK	(SPI_STAT_RXORERR | SPI_STAT_CONFERR | SPI_STAT_CRCERR)
+
+#define GD32_SPI_PSC_MAX	0x7U
+
+/** Obtain RCU register offset from RCU clock value */
+#define RCU_CLOCK_OFFSET(rcu_clock) ((rcu_clock) >> 6U)
+
+struct spi_gd32_config {
+	uint32_t reg;
+	uint32_t rcu_periph_clock;
+	const struct pinctrl_dev_config *pcfg;
+};
+
+struct spi_gd32_data {
+	struct spi_context ctx;
+};
+
+static int spi_gd32_get_err(const struct spi_gd32_config *cfg)
+{
+	uint32_t stat = SPI_STAT(cfg->reg);
+
+	if (stat & SPI_GD32_ERR_MASK) {
+		LOG_ERR("spi%u error status detected, err = %u",
+			cfg->reg, stat & (uint32_t)SPI_GD32_ERR_MASK);
+
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static bool spi_gd32_transfer_ongoing(struct spi_gd32_data *data)
+{
+	return spi_context_tx_on(&data->ctx) ||
+	       spi_context_rx_on(&data->ctx);
+}
+
+static int spi_gd32_configure(const struct device *dev,
+			      const struct spi_config *config)
+{
+	struct spi_gd32_data *data = dev->data;
+	const struct spi_gd32_config *cfg = dev->config;
+	uint32_t bus_freq;
+
+	if (spi_context_configured(&data->ctx, config)) {
+		return 0;
+	}
+
+	if (SPI_OP_MODE_GET(config->operation) == SPI_OP_MODE_SLAVE) {
+		LOG_ERR("Slave mode not supported");
+		return -ENOTSUP;
+	}
+
+	/* Before configure, disable spi first. */
+	SPI_CTL0(cfg->reg) &= ~SPI_CTL0_SPIEN;
+
+	/* Default to master mode. */
+	SPI_CTL0(cfg->reg) |= SPI_MASTER;
+	/* Default to full duplex transfer mode. */
+	SPI_CTL0(cfg->reg) &= ~SPI_TRANSMODE_BDTRANSMIT;
+
+	if (SPI_WORD_SIZE_GET(config->operation) == 8) {
+		SPI_CTL0(cfg->reg) |= SPI_FRAMESIZE_8BIT;
+	} else {
+		SPI_CTL0(cfg->reg) |= SPI_FRAMESIZE_16BIT;
+	}
+
+	/* Reset to hardware NSS mode. */
+	SPI_CTL0(cfg->reg) &= ~SPI_CTL0_SWNSSEN;
+	if (config->cs != NULL) {
+		SPI_CTL0(cfg->reg) |= SPI_CTL0_SWNSSEN;
+	} else {
+		/*
+		 * For signle master env,
```

Comment:
nit: single

Reply:
Thanks! I'm start to using code spell checker , may help to avoid this miss spell issue.